doc/source/notebooks/manual/tailor/mixture_density_network.ipynb

```diff
@@ -4,15 +4,745 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Mixture Density Models\n",
+    "\n",
+    "Mixture density networks (MDNs) are a parametric class of models that allow for conditional density estimation. They are built out of 2 parts: a neural net and a Mixture of Gaussians (MoG). The neural net is  responsible for producing the characteristics of the MoG. In practice, given that the MoG consists of $M$ Gaussians, the neural net will output a collection of $M$ means, variances and weights $\\{\\mu_m, \\sigma_m^2, \\pi_m\\}_{m=1}^M$. The produced means, variances and weights are used to define the conditional probability distribution function \n",
+    "$$\n",
+    "p(Y = y\\,|\\,X = x) = \\sum_{m=1}^{M} \\pi_{m}(x)\\,\\mathcal{N}\\big(y\\, \\left|\\,\\mu_{m}(x), \\sigma_{m}^2(x)\\big)\\right.\n",
+    "$$\n",
+    "Each of the parameters  $\\pi_{m}(x), \\mu_{m}(x), \\sigma_{m}(x)$ of the distribution will be determined by the neural network, as a function of the input $x$.\n",
+    "\n",
+    "We train the MDN's neural net by optimizing the model's likelihood\n",
+    "$$\n",
+    " \\mathcal{L} \\triangleq \\operatorname{argmax}_{\\Theta} \\prod_{n=1}^N p(Y = y_n\\,|\\,X = x_n),\n",
+    "$$\n",
+    "where $\\Theta$ collects the neural nets weights and biases and $\\{x_n, y_n\\}_{n=1}^N$ are our training dataset."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## GPflow MDN implementation\n",
+    "\n",
+    "GPflow doesn't reinvent the wheel: most of what you will see is just plain Python/TensorFlow code. We choose to use GPflow, however, because it provides us with some functionality to easily define a model. Once we have a GPflow model, we can specify its objective function, parameters, dataset. This extra layer of abstraction makes interacting with the model, such as optimizing or performing inference, much easier. \n",
+    "\n",
+    "We begin by importing the required packages from TensorFlow and GPflow"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import tensorflow as tf\n",
+    "from scipy.stats import norm"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import gpflow\n",
+    "from gpflow.models.model import Model\n",
+    "from gpflow.params.dataholders import Minibatch, DataHolder\n",
+    "from gpflow.params import Parameter, ParamList\n",
+    "from gpflow.training import AdamOptimizer, ScipyOptimizer\n",
+    "from gpflow.decors import params_as_tensors, autoflow\n",
```

Comment:
`params_as_tensors` and `autoflow` is available though `gpflow`, pls, do not use `decors`